node/bin/src/prover_api/snark_job_manager.rs

```diff
@@ -60,73 +40,56 @@ pub struct SnarkJobManager {
         // config
         max_fris_per_snark: usize,
+        assignment_timeout: Duration,
+        max_assigned_batch_range: usize,
     ) -> Self {
+        let jobs = ProverJobMap::<FriProof>::new(
+            assignment_timeout,
+            max_assigned_batch_range,
+            ProverStage::Snark,
+        );
         let latency_tracker = ComponentStateReporter::global().handle_for(
             "snark_job_manager",
             GenericComponentState::ProcessingOrWaitingRecv,
         );
-        let committed_batch_receiver = Mutex::new(committed_batch_receiver);
         Self {
-            committed_batch_receiver,
+            jobs,
             prove_batches_sender,
             max_fris_per_snark,
             latency_tracker,
         }
     }
 
+    /// Adds a pending job to the queue.
+    /// Awaits if queue is full (ProverJobMap.max_assigned_batch_range).
+    pub async fn add_job(&self, batch_envelope: SignedBatchEnvelope<FriProof>) {
+        self.jobs.add_job(batch_envelope).await
+    }
+
     // If there is a job pending, returns a non-empty list of tuples (`batch_number`, `verification_key_hash`, `real_fri_proof`)
-    pub async fn pick_real_job(&self) -> anyhow::Result<Option<Vec<(FriJob, FriProof)>>> {
-        self.consume_fake_proves_from_head(None).await?;
-        // note that here we don't consume the messages from channel -
-        // the job will be picked, but there is no guarantee it will be completed
-        let batches_with_real_proofs: Vec<(FriJob, FriProof)> = self
-            .committed_batch_receiver
-            .lock()
-            .await
-            .peek_until(self.max_fris_per_snark, |envelope| {
-                if envelope.data.is_fake() {
-                    None
-                } else {
-                    let proving_execution_version = ProvingVersion::try_from(
-                        envelope
-                            .data
-                            .proving_execution_version()
-                            .expect("proving execution version must be present on proof"),
-                    )
-                    .expect("execution version must exist as it was set by server");
-                    Some((
-                        FriJob {
-                            batch_number: envelope.batch_number(),
-                            vk_hash: proving_execution_version.vk_hash().to_string(),
-                        },
-                        envelope.data.clone(),
-                    ))
-                }
-            });
+    pub async fn pick_real_job(
+        &self,
+        prover_id: String,
+    ) -> anyhow::Result<Option<Vec<(FriJob, FriProof)>>> {
+        // consume/remove all fake jobs that may be in the front of the queue
+        self.process_pending_fake_fri_proofs().await?;
+
+        let batches_with_real_proofs = self
+            .jobs
+            .pick_jobs_while(self.max_fris_per_snark, &prover_id, |job| {
```

Comment:
We need to make sure that these batches have the same VK/proving_execution_version or otherwise we cannot seal such batch. I believe this is already done at batcher level, but I'd really love to make sure we double check here (i.e. provers can't merge/SNARKify 2 proofs with different VKs).